# Importing Libraries

In [1]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install folium

     |████████████████████████████████| 93 kB 3.2 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
                                                                                                                           \failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - cffi -> python[version='2.7.*|3.5.*|3.6.*|3.6.12|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|3.6.9|3.6.9|3.6.9|>=2.7,<2.8.0a0|3.6.9|>=3.5,<3.6.0a0|3.4.*',build='1_73_pypy|2_73_pypy|4_73_pypy|3_73_pypy|0_73_pypy']
  - geopy -> python[version='2.7.*|3.4.*|3.5.*|3.6.*']
  - rsa -> pyt

matplotlib -> matplotlib-base[version='2.1.2|2.1.2|2.1.2|2.2.3|2.2.3|2.2.3|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|2.2.4|3.0.1|3.0.2|3.0.2|3.0.2|3.0.2|3.0.2|3.0.3|3.0.3|3.0.3|3.0.3|3.1.0|3.1.0|3.1.0|3.1.0|3.1.1|3.1.1|3.1.1|3.1.1|3.1.1|3.1.1|3.1.1|3.1.1|3.1.2|3.1.2|3.1.2|3.1.2|3.1.2|3.1.2|3.1.3|3.1.3|3.1.3|>=3.2.0,<3.2.1.0a0|>=3.2.1,<3.2.2.0a0|>=3.2.2,<3.2.3.0a0|>=3.3.0,<3.3.1.0a0|>=3.3.1,<3.3.2.0a0|>=3.3.2,<3.3.3.0a0|>=3.3.3,<3.3.4.0a0|>=2.2.5,<2.2.6.0a0|3.1.3|3.1.3|3.1.3|3.1.2|3.1.2|3.1.2',build='py37hef1b27d_1|py37h250f245_1|py27h250f245_1|py36h250f245_1|py36h60b886d_1|py27hfd891ef_0|py36h250f245_1|py27h250f245_1|py37h250f245_2|py36h250f245_2|py37hc2fb212_1001|py36h167e16e_1000|py36h167e16e_1002|py37h5f35d83_1|py36h5f35d83_1|py37h5f35d83_0|py37hfd891ef_0|py37h250f245_2|py36h250f245_2|py37h250f245_0|py38h250f245_0|py36h250f245_1|py37h250f245_0|py38h250f245_0|py36h250f245_0|py38h250f245_1|py37h250f245_1|py36h250f245_0|py38h250f245_2|py38h250f245_1|py36he7580a8_1|py37he758

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: | 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
                                                                                                                                                 -failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - cffi -> python[version='2.7.*|3.5.*|3.6.*|3.6.12|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|3.6.9|3.6.9|3.6.9|>=2.7,<2.8.0a0|3.6.9|>=3.5,<3.6.0a0|3.4.*',build='1_73_pypy|4_73_pypy|3_73_pypy|2_73_pypy|0_73_pypy']
  - rsa -> python[version='2.7.*|3.4.*|3.5.*|3.6.*']

Your python: python=3.7

If python is on the left-most side of the chain, that's the version you've a

Libraries imported.


# Downloading Data in Dataframe

In [4]:
from bs4 import BeautifulSoup

In [5]:
url= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

source= requests.get(url).text
soup= BeautifulSoup(source)
table_data=soup.find('div',
class_='mw-parser-output')
table=table_data.table.tbody

columns= ['Postal Code', 'Borough', 'Neighborhood']
data=dict({key:[]*len(columns) for key in columns})

for row in table.find_all('tr'):
    for i, column in zip(row.find_all('td'),columns):
        i=i.text
        i=i.replace('\n',
'')
        data[column].append(i)
        
df= pd.DataFrame.from_dict(data=data)[columns]
print(df.shape)
df.head()

(180, 3)


Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

# Dropping Not Assigned Boroughs

In [6]:
df= df[df['Borough'] != 'Not assigned'].reset_index(drop = True)
print('After dropping rows where borough is "Not Assigned", Shape is: ',df.shape)
print('Number of rows where Neighborhood is "Not Assigned" but borough has value: ',
     df[df['Neighborhood'] == 'Not assigned'].shape[0])

After dropping rows where borough is "Not Assigned", Shape is:  (103, 3)
Number of rows where Neighborhood is "Not Assigned" but borough has value:  0


In [7]:
p, b, n =[], [], []

for postcode, borough, neigh in zip(df['Postal Code'], df['Borough'], df['Neighborhood']):
    p.append(postcode)
    b.append(borough)
    if neigh == 'Not assigned':
        n.append(borough)
    else:
        n.append(neigh)

df=pd.DataFrame({'Postal Code':p,'Borough':b, 'Neighborhood':n})[columns]
print(df.shape)
df.head()

(103, 3)


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

# <font size="4"> 77 rows have been dropped because there was no assigned boroughs for them.</font>

# <font size="5"> Original Data had 180 rows and now the data is having 103 rows.</font>

In [8]:
postcodes=df['Postal Code'].values
boroughs =df['Borough'].values
neighs=df['Neighborhood'].values

dic=dict({(key1, key2): [] for key1, key2 in zip(postcodes, boroughs)})
print('Number of keys in the dictionary are: ',len(dic.keys()))

for postcode, borough, neigh in zip(postcodes, boroughs, neighs):
    key=(postcode, borough)
    dic[key].append(neigh)

df=pd.DataFrame(columns=['Postal Code','Borough','Neighborhood'])
for key, value in dic.items():
    postcode, borough, neig=key[0],key[1], value
    neig=', '.join(neig)
    df=df.append({'Postal Code': postcode,
                 'Borough': borough,
                 'Neighborhood': neig}, ignore_index=True)

print('Shape of final data is: ',df.shape)
df.head(10)

Number of keys in the dictionary are:  103
Shape of final data is:  (103, 3)


Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

# Latitude and Longitude of Each Neighbourhood

In [9]:
pip install geocoder

     |████████████████████████████████| 98 kB 8.5 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [10]:
pip install pgeocode

Note: you may need to restart the kernel to use updated packages.


In [11]:
import geocoder # import geocoder

# Downlaoding the GeoSpatial Data by downloading coordinates of postal codes

In [12]:
!wget http://cocl.us/Geospatial_data

--2020-12-20 12:45:32--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 169.63.96.176, 169.63.96.194
Connecting to cocl.us (cocl.us)|169.63.96.176|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2020-12-20 12:45:32--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|169.63.96.176|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-12-20 12:45:33--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 185.235.236.197
Connecting to ibm.box.com (ibm.box.com)|185.235.236.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-12-20 12:45:33--  https://ibm.box.com/public/static/9afzr8

In [13]:
try:
    df['Latitude']= lats
    df['Longitude']= lons
except:
    latlon=pd.read_csv('Geospatial_data')
    df=pd.merge(df, latlon, how='inner', on= 'Postal Code')

print(df.shape)
df.head(10)

(103, 5)


Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

# Exploring and Clustering Toronto Neighborhoods

In [14]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [15]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [16]:
print('Total number of Borough= ', len(df['Borough'].unique()))

Total number of Borough=  10


# Exploring one borough for reference at the start and then we will check other neighbourhoods as well

In [18]:
NYORK_toronto=df[df['Borough']=='North York'].reset_index(drop=True)
print(NYORK_toronto.shape)
NYORK_toronto.head()

(24, 5)


Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.464763  
3 -79.352188  
4 -79.445073

# Visualized Map of Downtown Toronto

In [19]:
address = 'North York, Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Downtown Toronto using latitude and longitude values
map_NYORK = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(downtown_toronto['Latitude'], NYORK_toronto['Longitude'], NYORK_toronto['Borough'],NYORK_toronto['Neighborhood']):
    label='{},{}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_NYORK)  
    
map_NYORK

# Exploring Venues using Foursquare API

In [20]:
lat= NYORK_toronto.loc[0,'Latitude']
lon= NYORK_toronto.loc[0,'Longitude']
neighborhood_name= NYORK_toronto.loc[0,'Neighborhood']

print('Latitude and Longitude of the neighborhood as {} are {}, {} '.format(neighborhood_name, lat,lon))

Latitude and Longitude of the neighborhood as Parkwoods are 43.7532586, -79.3296565 


In [21]:
CLIENT_ID = 'C11DJYAJWWUVQFE4UHBQ5A24ZKZBHVMHYJQLJ4MOK5H1YL1J' # your Foursquare ID
CLIENT_SECRET = '3IPTYS4WO4AAGN1FLIDQVL5O3SLSOS4TSJTLBWDEC4Y22OQN' # your Foursquare Secret
VERSION = '20201220' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

In [22]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    lat, 
    lon, 
    radius, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fdf493d499ac81c60028841'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 28,
  'suggestedBounds': {'ne': {'lat': 43.762258609000014,
    'lng': -79.31721997969855},
   'sw': {'lat': 43.74425859099999, 'lng': -79.34209302030145}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b8991cbf964a520814232e3',
       'name': "Allwyn's Bakery",
       'location': {'address': '81 Underhill drive',
        'lat': 43.75984035203157,
        'lng': -79.32471879917513,
        'labeledLatLngs': [{'label': 'display'

In [23]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name            categories        lat        lng
0     Allwyn's Bakery  Caribbean Restaurant  43.759840 -79.324719
1         Tim Hortons                  Café  43.760668 -79.326368
2     Brookbanks Park                  Park  43.751976 -79.332140
3   Bruno's valu-mart         Grocery Store  43.746143 -79.324630
4  Shoppers Drug Mart              Pharmacy  43.760857 -79.324961

# Venues and Details for All Neighborhoods of Toronto

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [26]:
NYORK_venues = getNearbyVenues(names=NYORK_toronto['Neighborhood'],
                                   latitudes=NYORK_toronto['Latitude'],
                                   longitudes=NYORK_toronto['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Manor, Lawrence Heights
Don Mills
Glencairn
Don Mills
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Fairview, Henry Farm, Oriole
Northwood Park, York University
Bayview Village
Downsview
York Mills, Silver Hills
Downsview
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Willowdale, Newtonbrook
Downsview
Bedford Park, Lawrence Manor East
Humberlea, Emery
Willowdale, Willowdale East
Downsview
York Mills West
Willowdale, Willowdale West


In [27]:
print(NYORK_venues.shape)
NYORK_venues.head()

(250, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1               Variety Store       43.751974       -79.333114   
2  TTC stop - 44 Valley Woods       43.755402       -79.333741   
3      Victoria Village Arena       43.723481       -79.315635   
4                 Tim Hortons       43.725517       -79.313103   

      Venue Category  
0               Park  
1  Food & Drink Shop  
2           Bus Stop  
3       Hockey Arena  
4        Coffee Shop

In [28]:
NYORK_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North                     23   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
Don Mills                                                           25   
Downsview                                                           14   
Fairview, Henry Farm, Oriole                                        70   
Glencairn                                                            4   
Hillcrest Village                                                    5   
Humber Summit                                                        3   
Humberlea, Emery                                                     1   
Lawrence Manor, Lawrence Heights                                    14   
North Park, Maple Leaf Park, Upwood Park                             5   
Northwood Park, York University                                      7   
Parkwoods                                                            3   
Victoria Village                                                     4   
Willowdale, Newtonbrook                                              1   
Willowdale, Willowdale East                                         34   
Willowdale, Willowdale West                                          6   
York Mills West                                                      3   
York Mills, Silver Hills                                             2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Wilson Heights, Downsview North                      23   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
Don Mills                                                            25   
Downsview                                                            14   
Fairview, Henry Farm, Oriole                                         70   
Glencairn                                                             4   
Hillcrest Village                                                     5   
Humber Summit                                                         3   
Humberlea, Emery                                                      1   
Lawrence Manor, Lawrence Heights                                     14   
North Park, Maple Leaf Park, Upwood Park                              5   
Northwood Park, York University                                       7   
Parkwoods                                                             3   
Victoria Village                                                      4   
Willowdale, Newtonbrook                                               1   
Willowdale, Willowdale East                                          34   
Willowdale, Willowdale West                                           6   
York Mills West                                                       3   
York Mills, Silver Hills                                              2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North     23              23   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
Don Mills                                           25              25   
Downsview                                           14              14   
Fairview, Henry Farm, Oriole                        70              70   
Glencairn                                            4               4   
Hillcrest Village                                    5               5   
Humber Summit                

In [29]:
print('There are {} unique categories.'.format(len(NYORK_venues['Venue Category'].unique())))

There are 105 unique categories.


# Analyzing each neighborhood

In [30]:
# one hot encoding
NYORK_onehot = pd.get_dummies(NYORK_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
NYORK_onehot['Neighborhood'] = NYORK_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [NYORK_onehot.columns[-1]] + list(NYORK_onehot.columns[:-1])
NYORK_onehot = NYORK_onehot[fixed_columns]

NYORK_onehot.head()

Neighborhood  Accessories Store  Airport  American Restaurant  \
0         Parkwoods                  0        0                    0   
1         Parkwoods                  0        0                    0   
2         Parkwoods                  0        0                    0   
3  Victoria Village                  0        0                    0   
4  Victoria Village                  0        0                    0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   Bakery  Bank  Bar  Baseball Field  Basketball Court  Beer Store  Bike Shop  \
0       0     0    0               0                 0           0          0   
1       0     0    0               0                 0           0          0   
2       0     0    0               0                 0           0          0   
3       0     0    0               0                 0           0          0   
4       0     0    0               0                 0           0          0   

   Boutique  Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  \
0         0            0                0             0              0   
1         0            0                0             0              0   
2         0            0                0             0              0   
3         0            0                0             0              0   
4         0            0                0             0              0   

   Bus Stop  Business Service  Butcher  Cafeteria  Café  Caribbean Restaurant  \
0         0                 0        0          0     0                     0   
1         0                 0        0          0     0                     0   
2         1                 0        0          0     0                     0   
3         0                 0        0          0     0                     0   
4         0                 0        0          0     0                     0   

   Chinese Restaurant  Chocolate Shop  Clothing Store  Coffee Shop  \
0                   0               0               0            0   
1                   0               0               0            0   
2                   0               0               0            0   
3                   0               0               0            0   
4                   0               0               0            1   

   Comfort Food Restaurant  Construction & Landscaping  Convenience Store  \
0                        0                           0                  0   
1                        0                           0                  0   
2                        0                           0                  0   
3                        0                           0                  0   
4                        0                           0                  0   

   Cosmetics Shop  Deli / Bodega  Department Store  Dessert Shop  \
0               0              0                 0             0   
1               0              0                 0             0   
2               0              0                 0             0   
3               0              0                 0             0   
4               0              0                 0             0   

   Dim Sum Restaurant  Diner  Discount Store  Distribution Center  Dog Run  \
0                   0      0               0                    0        0   
1                   0      0               0                    0        0   
2                   0      0               0                    0        0   
3                   0      0               0                    0        0   
4        

In [31]:
NYORK_onehot.shape

(250, 106)

In [33]:
NYORK_grouped = NYORK_onehot.groupby('Neighborhood').mean().reset_index()
NYORK_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                   Bayview Village           0.000000   
2                 Bedford Park, Lawrence Manor East           0.000000   
3                                         Don Mills           0.000000   
4                                         Downsview           0.000000   
5                      Fairview, Henry Farm, Oriole           0.000000   
6                                         Glencairn           0.000000   
7                                 Hillcrest Village           0.000000   
8                                     Humber Summit           0.000000   
9                                  Humberlea, Emery           0.000000   
10                 Lawrence Manor, Lawrence Heights           0.071429   
11         North Park, Maple Leaf Park, Upwood Park           0.000000   
12                  Northwood Park, York University           0.000000   
13                                        Parkwoods           0.000000   
14                                 Victoria Village           0.000000   
15                          Willowdale, Newtonbrook           0.000000   
16                      Willowdale, Willowdale East           0.000000   
17                      Willowdale, Willowdale West           0.000000   
18                                  York Mills West           0.000000   
19                         York Mills, Silver Hills           0.000000   

     Airport  American Restaurant  Art Gallery  Arts & Crafts Store  \
0   0.000000             0.000000         0.00             0.000000   
1   0.000000             0.000000         0.00             0.000000   
2   0.000000             0.045455         0.00             0.000000   
3   0.000000             0.000000         0.04             0.000000   
4   0.071429             0.000000         0.00             0.000000   
5   0.000000             0.014286         0.00             0.000000   
6   0.000000             0.000000         0.00             0.000000   
7   0.000000             0.000000         0.00             0.000000   
8   0.000000             0.000000         0.00             0.000000   
9   0.000000             0.000000         0.00             0.000000   
10  0.000000             0.000000         0.00             0.000000   
11  0.000000             0.000000         0.00             0.000000   
12  0.000000             0.000000         0.00             0.000000   
13  0.000000             0.000000         0.00             0.000000   
14  0.000000             0.000000         0.00             0.000000   
15  0.000000             0.000000         0.00             0.000000   
16  0.000000             0.000000         0.00             0.029412   
17  0.000000             0.000000         0.00             0.000000   
18  0.000000             0.000000         0.00             0.000000   
19  0.000000             0.000000         0.00             0.000000   

    Asian Restaurant  Athletics & Sports    Bakery      Bank       Bar  \
0           0.000000            0.000000  0.000000  0.086957  0.000000   
1           0.000000            0.000000  0.000000  0.250000  0.000000   
2           0.000000            0.000000  0.000000  0.000000  0.000000   
3           0.000000            0.040000  0.000000  0.000000  0.000000   
4           0.000000            0.071429  0.000000  0.071429  0.000000   
5           0.014286            0.000000  0.028571  0.028571  0.014286   
6           0.000000            0.000000  0.250000  0.000000  0.000000   
7           0.000000            0.200000  0.000000  0.000000  0.000000   
8           0.000000            0.000000  0.000000  0.000000  0.000000   
9           0.000000            0.000000  0.000000  0.000000  0.000000   
10          0.000000            0.000000  0.000000  0.000000  0.000000   
11          0.000000            0.000000  0.200000  0.000000  0.000000   
12          0.000000          

In [34]:
NYORK_grouped.shape

(20, 106)

# 5 Most Common Venues of Each Neighborhood

In [35]:
num_top_venues = 5

for hood in NYORK_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = NYORK_grouped[NYORK_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor, Wilson Heights, Downsview North----
            venue  freq
0     Coffee Shop  0.09
1            Bank  0.09
2  Sandwich Place  0.04
3       Gift Shop  0.04
4  Ice Cream Shop  0.04


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1  Italian Restaurant  0.09
2         Coffee Shop  0.09
3            Pharmacy  0.05
4   Indian Restaurant  0.05


----Don Mills----
                 venue  freq
0                  Gym  0.12
1           Beer Store  0.08
2           Restaurant  0.08
3          Coffee Shop  0.08
4  Japanese Restaurant  0.08


----Downsview----
           venue  freq
0           Park  0.14
1  Grocery Store  0.14
2           Bank  0.07
3   Liquor Store  0.07
4  Shopping Mall  0.07


----Fairview, Henry Farm, Oriole--

# Checking 10 Most Common Venue of Each Neighborhood

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = NYORK_grouped['Neighborhood']

for ind in np.arange(NYORK_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(NYORK_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                                  Bayview Village    Chinese Restaurant   
2                Bedford Park, Lawrence Manor East        Sandwich Place   
3                                        Don Mills                   Gym   
4                                        Downsview         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank           Gas Station        Sandwich Place   
1                  Café                  Bank   Japanese Restaurant   
2           Coffee Shop    Italian Restaurant         Grocery Store   
3            Beer Store   Japanese Restaurant            Restaurant   
4                  Park  Gym / Fitness Center               Airport   

       5th Most Common Venue    6th Most Common Venue  \
0  Middle Eastern Restaurant           Ice Cream Shop   
1             Discount Store  Comfort Food Restaurant   
2                  Juice Bar                     Café   
3                Coffee Shop                     Café   
4                 Food Truck           Discount Store   

        7th Most Common Venue    8th Most Common Venue 9th Most Common Venue  \
0           Mobile Phone Shop                     Park         Deli / Bodega   
1  Construction & Landscaping        Convenience Store        Cosmetics Shop   
2                     Butcher  Comfort Food Restaurant     Indian Restaurant   
3              Baseball Field       Dim Sum Restaurant        Discount Store   
4                Liquor Store         Business Service    Athletics & Sports   

  10th Most Common Venue  
0               Pharmacy  
1          Deli / Bodega  
2               Pharmacy  
3              Bike Shop  
4                   Bank

# Clustering Neighbours

In [38]:
# set number of clusters
kclusters = 5

NYORK_grouped_clustering = NYORK_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(NYORK_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 4, 0, 4, 0, 4, 3, 2], dtype=int32)

In [39]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels_1', kmeans.labels_)

NYORK_merged = NYORK_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
NYORK_merged = NYORK_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

NYORK_merged.head() # check the last columns!

Postal Code     Borough                      Neighborhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  Cluster Labels_1 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656                 0                  Park     Food & Drink Shop   
1 -79.315572                 4     French Restaurant           Coffee Shop   
2 -79.464763                 4        Clothing Store     Accessories Store   
3 -79.352188                 4                   Gym            Beer Store   
4 -79.445073                 0                  Park                Bakery   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0              Bus Stop          Women's Store                 Diner   
1          Hockey Arena  Portuguese Restaurant         Women's Store   
2            Shoe Store  Vietnamese Restaurant           Event Space   
3   Japanese Restaurant             Restaurant           Coffee Shop   
4                   Pub    Japanese Restaurant         Women's Store   

  6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
0           Coffee Shop  Comfort Food Restaurant  Construction & Landscaping   
1    Dim Sum Restaurant           Clothing Store     Comfort Food Restaurant   
2           Coffee Shop       Miscellaneous Shop                    Boutique   
3                  Café           Baseball Field          Dim Sum Restaurant   
4    Dim Sum Restaurant              Coffee Shop     Comfort Food Restaurant   

        9th Most Common Venue 10th Most Common Venue  
0           Convenience Store         Cosmetics Shop  
1  Construction & Landscaping      Convenience Store  
2      Furniture / Home Store                   Bank  
3              Discount Store              Bike Shop  
4  Construction & Landscaping      Convenience Store

In [41]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(NYORK_merged['Latitude'], NYORK_merged['Longitude'], NYORK_merged['Neighborhood'], NYORK_merged['Cluster Labels_1']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examining Clusters

### Cluster 1

In [42]:
NYORK_merged.loc[NYORK_merged['Cluster Labels_1'] == 0, NYORK_merged.columns[[1] + list(range(5, NYORK_merged.shape[1]))]]

Borough  Cluster Labels_1 1st Most Common Venue 2nd Most Common Venue  \
0   North York                 0                  Park     Food & Drink Shop   
4   North York                 0                  Park                Bakery   
11  North York                 0         Grocery Store                  Park   
13  North York                 0         Grocery Store                  Park   
14  North York                 0                 Trail                  Park   
17  North York                 0         Grocery Store                  Park   
21  North York                 0         Grocery Store                  Park   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                     Bus Stop         Women's Store                 Diner   
4                          Pub   Japanese Restaurant         Women's Store   
11        Gym / Fitness Center               Airport            Food Truck   
13        Gym / Fitness Center               Airport            Food Truck   
14  Construction & Landscaping                Bakery      Basketball Court   
17        Gym / Fitness Center               Airport            Food Truck   
21        Gym / Fitness Center               Airport            Food Truck   

   6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
0            Coffee Shop  Comfort Food Restaurant  Construction & Landscaping   
4     Dim Sum Restaurant              Coffee Shop     Comfort Food Restaurant   
11        Discount Store             Liquor Store            Business Service   
13        Discount Store             Liquor Store            Business Service   
14         Women's Store                    Diner                 Coffee Shop   
17        Discount Store             Liquor Store            Business Service   
21        Discount Store             Liquor Store            Business Service   

         9th Most Common Venue 10th Most Common Venue  
0            Convenience Store         Cosmetics Shop  
4   Construction & Landscaping      Convenience Store  
11          Athletics & Sports                   Bank  
13          Athletics & Sports                   Bank  
14     Comfort Food Restaurant      Convenience Store  
17          Athletics & Sports                   Bank  
21          Athletics & Sports                   Bank

### Cluster 2

In [43]:
NYORK_merged.loc[NYORK_merged['Cluster Labels_1'] == 1, NYORK_merged.columns[[1] + list(range(5, NYORK_merged.shape[1]))]]

Borough  Cluster Labels_1 1st Most Common Venue 2nd Most Common Venue  \
12  North York                 1                  Park             Cafeteria   
16  North York                 1                  Park         Women's Store   
22  North York                 1                  Park     Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12         Women's Store                 Diner           Coffee Shop   
16                 Diner        Clothing Store           Coffee Shop   
22         Women's Store                 Diner        Clothing Store   

      6th Most Common Venue       7th Most Common Venue  \
12  Comfort Food Restaurant  Construction & Landscaping   
16  Comfort Food Restaurant  Construction & Landscaping   
22              Coffee Shop     Comfort Food Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12           Convenience Store        Cosmetics Shop          Deli / Bodega  
16           Convenience Store        Cosmetics Shop          Deli / Bodega  
22  Construction & Landscaping        Cosmetics Shop          Deli / Bodega

### Cluster 3

In [44]:
NYORK_merged.loc[NYORK_merged['Cluster Labels_1'] == 2, NYORK_merged.columns[[1] + list(range(5, NYORK_merged.shape[1]))]]

Borough  Cluster Labels_1 1st Most Common Venue 2nd Most Common Venue  \
19  North York                 2        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue    5th Most Common Venue  \
19        Discount Store           Coffee Shop  Comfort Food Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19  Construction & Landscaping     Convenience Store        Cosmetics Shop   

   9th Most Common Venue 10th Most Common Venue  
19         Deli / Bodega       Department Store

### Cluster 4

In [45]:
NYORK_merged.loc[NYORK_merged['Cluster Labels_1'] == 3, NYORK_merged.columns[[1] + list(range(5, NYORK_merged.shape[1]))]]

Borough  Cluster Labels_1   1st Most Common Venue  \
15  North York                 3  Furniture / Home Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
15           Pizza Place                   Gym        Discount Store   

   5th Most Common Venue    6th Most Common Venue       7th Most Common Venue  \
15           Coffee Shop  Comfort Food Restaurant  Construction & Landscaping   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
15     Convenience Store        Cosmetics Shop          Deli / Bodega

### Cluster 5

In [46]:
NYORK_merged.loc[NYORK_merged['Cluster Labels_1'] == 4, NYORK_merged.columns[[1] + list(range(5, NYORK_merged.shape[1]))]]

Borough  Cluster Labels_1     1st Most Common Venue  \
1   North York                 4         French Restaurant   
2   North York                 4            Clothing Store   
3   North York                 4                       Gym   
5   North York                 4                       Gym   
6   North York                 4  Mediterranean Restaurant   
7   North York                 4               Coffee Shop   
8   North York                 4            Clothing Store   
9   North York                 4    Furniture / Home Store   
10  North York                 4        Chinese Restaurant   
18  North York                 4            Sandwich Place   
20  North York                 4          Ramen Restaurant   
23  North York                 4               Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1            Coffee Shop          Hockey Arena  Portuguese Restaurant   
2      Accessories Store            Shoe Store  Vietnamese Restaurant   
3             Beer Store   Japanese Restaurant             Restaurant   
5             Beer Store   Japanese Restaurant             Restaurant   
6            Golf Course    Athletics & Sports                   Pool   
7                   Bank           Gas Station         Sandwich Place   
8            Coffee Shop  Fast Food Restaurant             Restaurant   
9            Coffee Shop  Caribbean Restaurant         Massage Studio   
10                  Café                  Bank    Japanese Restaurant   
18           Coffee Shop    Italian Restaurant          Grocery Store   
20            Restaurant                  Café            Pizza Place   
23               Butcher              Pharmacy          Grocery Store   

        5th Most Common Venue    6th Most Common Venue  \
1               Women's Store       Dim Sum Restaurant   
2                 Event Space              Coffee Shop   
3                 Coffee Shop                     Café   
5                 Coffee Shop                     Café   
6                     Dog Run            Women's Store   
7   Middle Eastern Restaurant           Ice Cream Shop   
8         Japanese Restaurant                   Bakery   
9          Falafel Restaurant                      Bar   
10             Discount Store  Comfort Food Restaurant   
18                  Juice Bar                     Café   
20             Sandwich Place              Coffee Shop   
23                Pizza Place           Discount Store   

         7th Most Common Venue    8th Most Common Venue  \
1               Clothing Store  Comfort Food Restaurant   
2           Miscellaneous Shop                 Boutique   
3               Baseball Field       Dim Sum Restaurant   
5               Baseball Field       Dim Sum Restaurant   
6           Dim Sum Restaurant              Coffee Shop   
7            Mobile Phone Shop                     Park   
8                   Shoe Store                Juice Bar   
9           Miscellaneous Shop       Dim Sum Restaurant   
10  Construction & Landscaping        Convenience Store   
18                     Butcher  Comfort Food Restaurant   
20               Shopping Mall            Movie Theater   
23               Women's Store       Dim Sum Restaurant   

         9th Most Common Venue      10th Most Common Venue  
1   Construction & Landscaping           Convenience Store  
2       Furniture / Home Store                        Bank  
3               Discount Store                   Bike Shop  
5               Discount Store                   Bike Shop  
6      Comfort Food Restaurant  Construction & Landscaping  
7                Deli / Bodega                    Pharmacy  
8          Sporting Goods Shop                        Bank  
9      Comfort Food Restaurant  Construction & Landscaping  
10              Cosmetics Shop               Deli / Bodega  
18           Indian Restaurant                    Pharmacy  
20                   Pet Store               Grocery Sto